### ISSUE TITLE
Add support for binned aggregation

### ISSUE THREAD

**Issue by willium** [vega/vega#2959]
# Proposal
Introduce a new transform or augment the existing aggregation transform to allow for specifying a binning definition.

# Why 
Because binning and aggregation are currently implemented in sequential transforms, it's currently not possible to—for example—have bins with count 0. Rather, those bins are dropped before they reach the aggregation transform.

Consider [this visualization](https://vega.github.io/editor/#/url/vega/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykSArJQBWENgDsQAGhBMkUANZkATmwbiAJmhAB3GHTjSQOJBo01xZNAJk6aG+mgBMABhcz4NMlmduZGpEwUVABtUHEkBEN0AKCjBmUAG21Y-gQ2Yho4CBExSRkAMzZlZEw0UEwATxxokFEJEABfGUxlJHEIIpK0MJAqmu04AA9MOHEywqzErXQASQBZABEAIQACACVAiysZCC8I5PQaBA0mAH1h0fGmqQrq2qYLIwKpmZAFlY2tyyMkCB6QMdTmcIEFlBNASdzmMtABdXb7JCHSHAx4dIyXMZlVCgPZkA7aIHQkZYprNO4DdCY66TODTbTrNjYMarAAqbFKbGyX0w2yMeIJ6FUzPEFxJ13JfXu2jRz1eDKZVzZHMCXIgPL5Mj+AOFVxBYIhurGF00IHhIAFSO0RtFaP+Mmp2NxiORNrFVzKjUl-VqSDIKj4oyMKjUOCYlQBRP1SHBRijMKMbtBMcNiuNCfNGlUODQBSREDgWv+oRAUDU11hjUr5uQygUxd6ESi2lr9aMPutcCgENBlUS0RCIGUXbK5oKqgQ5RAqRSgRQzRAYzLGmioAYOFiq5AL0Sh1xUCRtTLiWKcrpbzL6k9MiGTin0EP2iGZ-pRyh0djMjYBQKBexLkoAQF2fHELQPfsnxfN541NL8fz-NAAKAmRKjvUCHwg9AI1pV8hzTUVk0-EBv1-OB-0AhcI3Q8Damw7d5SFfCTS0ODSPIoCvSrXYaIbcJIlqZ8WmldBEgsOAYyMDQVSedCXW0EIoztShCMwKRVkUiwclBNgcFhG4h3aMgB3cZ18StdA7AcGAmnNO17zkt9UU0-SAC84FUXN8zgSUm1o9thJAUTxHE5RJOkyRZLM5EQjdJSVLU2LNOUzAdL0hc2ksAdTMFDA4C8HxmhcWznMinLEvRBc3I81A80SAsfP47Rj1PITKUCsSJP8cKpxnGI5ygpry2vAzMu0eBAmQHMZAsUZlBwNhEkCWoYCgZIZCqthPLq7zzSQIZsh6fdH3QQTiOULJrnQJhFQ5YNzpmWqCxaOhMPeJY1k2XlLFWAAKNFgo0ABKIxFqYOkADFEgYCBrNQVoGELQKkDBxIAHlSCSJAczh5QEeexQAGEhvsqKmryxIfss+gCAARhcYHKosFcQJpyUML8r9ztJES4AKCEVHsLanr6F7akZEVlU5blPu2X7-rgIGQeRul0fcxbsfhxHeUJ4nSvM0tyZ+zxvEwWn6dcpnhjQVnzX7IzNF4+jd0Gk9QpkFRTAuzAABkxjIRw9eRVbIhwI28pNtSADYABY1Ncc2F15TBXqJq9Vm-DYu2KDR-irRogA):

![image](https://user-images.githubusercontent.com/356976/98181396-fef13080-1eb7-11eb-8bae-cbee7b07f0d2.png)
It's not possible to start the color encodings scale at 0, like it is with the [heatmap transform](https://vega.github.io/editor/#/url/vega/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykSArJQBWENgDsQAGhBMkUANZkATmwbiAJmhAB3GHTjSQOJBo01xZNAJk6aG+mgBMABhcz4NMlmduZGpEwUVABtUHEkBEN0AKCjBmUAG21Y-gQ2Yho4CBExSRkAMzZlZEw0UEwATxxokFEJEABfRqlwyNqNOHEIOkqjMQSoDsCUGUxlJG6ikrQwkCqa7QVOpy0ZHoAvaLmesgjk9DsHGCbWkF397U9vMsaAXRkAD3KQOEecZW1oJES4AAoAOSPAFSAAEsQYCBCAIAkgBZAAiACFQQAlQIWMgAu4ASlOID6qFAbw+XygP3+AMqIPBgUh0NRbGwXVBABU2KU2Nk0RjLNi8S1ZJMNEdHKF3C4HiAholEgBlGhbNAAFhkUDU4kwEDQ4wYcBaFWqtXggWQOCMBSyiS06BU9iM6sSxReJM+6G+v0BnW6vRpEIQlDAxB+etBBFpmEhgeQjwFMjYJigvTQAEZ7vd1uTftrQm0otpnmMjdpEhY4EhPjJVOobbq4DJxDQhjrlHqZFtVC22yANBykBYXqkUiMLVabSB4cieZhMfiJpZaqKTgaQBF8+g+kXFuhS+Jy5WQNXNF366um9E6+24J3UJee32B0SeyOYiPCmPtIzmeI2X3MFyIGnWdBXnMhjTgLwfBXNcOi6HoqiMBZal3fcjA7NgT38B9JHFKQUylUDolALhEFqSo4FlNgdBUOAuiYRI9SaDMQCQR5slmYiszIoxe2QR8Ch+CBT2KLJNRLOACjKMY6F+AAFUxzEsaxpMwX5tDhGhs1BGplFBABxH4nXyEA2AKAohLKVAnGY34wM0HM5ktWUUjg5MVLU9AAGENSkkAVFMUTMAAGS6MgxWIrxLnQa4sFBABaUEnAADiYqVkGUBQHMNbc-Orc0ZC6dVOheBgcFiIjdHscLzkin5tCXfEYssiK9jq6KIJuJjBVY9jc3OLiCx47C0AExIhPjZRAu0JgmX-BBEKbBRvPUSybBARIkCYCiADEGIgE5b1bU8Z1U2pJxRdEZyU5j0syjj5mLdAaGQMCLVUeanyHGJXIQwVCrYYqn1K8qXmeJ9gwY6IXEFQlQAhxjUGhmRnqQV6n0tCjxxR17BQap8LjayrjkajqfHx2qDgwUnbhkJAIBqKBmpAeHolGoTmmaO50yAA): 
![image](https://user-images.githubusercontent.com/356976/98182178-fd286c80-1eb9-11eb-88ae-43d7ad587a4f.png)

This seems to be those bins are made to be null and dropped prior to reaching the aggregation transform and subsequently being `groupby`ed, and before it seems the `drop` property can be used to _not_ drop empty groups.

**Comment by jheer** [comment]
1. You can already make accurate color scales in Vega. Simply include zero (or the scale zero: true option) in the domain of the scale backing the color legend. Instead, I think what you are asking for is the ability to include zero-valued cells that have a corresponding color?
2. I'm not sure what hash map you are talking about. Do you mean heatmap?

Have you tried using the `cross` parameter of the [aggregate transform](https://vega.github.io/vega/docs/transforms/aggregate/)?

**Comment by domoritz** [comment]
Cross could work for 2D charts but not for 1D charts such as [this histogram](https://vega.github.io/editor/#/url/vega-lite/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykBaADZ04JACyUAVhDYA7EABoQAEySYUqUAwBOgtCrVICCNsRpwIUmfIC+SzNqSyIAMzbaEaANqgXNQZhw2vqqmAwIXgDkAJIAsgAiAEIABABKajSyZJEAusnJADzJAKzJAD5lyaHhUXFJaRlZufkAfMkA7CA2OUrI2gDW+tpwUJiKIHCyUGzKmWRooAAeCyBMmWgODHBKfnCCyvp1KemYc11K04LuK0hkZMMUgfrTDLJjNh9AA).

![image](https://user-images.githubusercontent.com/589034/98277258-0235fc00-1f4c-11eb-97df-4b3be1115064.png)

Even though we do have `cross`, I wonder whether a binned aggregation transform could be more efficient since it could a) use an array to store the aggregate values instead of needing a hash map and b) wouldn't need to create intermediate tuples with the bin start and bin end values.

**Comment by willium** [comment]
> 1. You can already make accurate color scales in Vega. Simply include zero (or the scale zero: true option) in the domain of the scale backing the color legend. Instead, I think what you are asking for is the ability to include zero-valued cells that have a corresponding color?

Yeah, essentially I'd like zero-valued cells (as outputted by a binning transform) to be mapped to zero on the color scale. The zero property [being true](https://vega.github.io/editor/#/url/vega/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykSArJQBWENgDsQAGhBMkUANZkATmwbiAJmhAB3GHTjSQOJBo01xZNAJk6aG+mgBMABhcz4NMlmduZGpEwUVABtUHEkBEN0AKCjBmUAG21Y-gQ2Yho4CBExSRkAMzZlZEw0UEwATxxokFEJEABfGUxlJHEIIpK0MJAqmu04AA9MOHEywqzErXQASQBZABEAIQACACVAiysZCC8I5PQaBA0mAH1h0fGmqQrq2qYLIwKpmZAFlY2tyyMkCB6QMdTmcIEFlBNASdzmMtABdXb7JCHSHAx4dIyXMZlVCgPZkA7aIHQkZYprNO4DdCY66TODTbTrNjYMarAAqbFKbGyX0w2yMeIJ6FUzPEFxJ13JfXu2jRz1eDKZVzZHMCXIgPL5Mj+AOFVxBYIhurGF00IHhIAFSO0RtFaP+Mmp2NxiORNrFVzKjUl-VqSDIKj4oyMKjUOCYlQBRP1SHBRijMKMbtBMcNiuNCfNGlUODQBSREDgWv+oRAUDU11hjUr5uQygUxd6ESi2lr9aMPutcCgENBlUS0RCIGUXbK5oKqgQ5RAqRSgRQzRAYzLGmioAYOFiq5AL0Sh1xUCRtTLiWKcrpbzL6k9MiGTin0EP2iGZ-pRyh0djMjYBQKBexLkoAQF2fHELQPfsnxfN541NL8fz-NAAKAmRKjvUCHwg9AI1pV8hzTUVk0-EBv1-OB-0AhcI3Q8Damw7d5SFfCTS0ODSPIoCvSrXYaIbcJIlqZ8WmldBEgsOAYyMDQVSedCXW0EIoztShCMwKRVkUiwclBNgcFhG4h3aMgB3cZ18StdA7AcGAmnNO17zkt9UU0-SAC84FUXN8zgSUm1o9thJAUTxHE5RJOkyRZLM5EQjdJSVLU2LNOUzAdL0hc2ksAdTMFDA4C8HxmhcWznMinLEvRBc3I81A80SAsfP47Rj1PITKUCsSJP8cKpxnGI5ygpry2vAzMu0eBAmQHMZAsUZlBwNhEkCWoYCgZIZCqtg0FaBhvPNJAhmyHp90fdBBOI5QsmudAmEVDlgwumZaoLFo6Ew94ljWTZeUsVYAAo0WCjQAEojEWpg6QAMUSBgIGs1BtsLQKkHBxIAHlSCSJAc3h5QdpexQAGEhvsqKmryxJfss+gCAARhcEHKosFcQJpyUML8r8LtJES4AKCEVHsTy6sR3lMDexkRWVTluS+7Y-oBuBgdB5G6XR9zFuxhH8YUImrxJnKoHJ37PG8TBafp1ymeGNBWfNfsjM0Xj6N3QaT1CmQVFMS7MAAGTGMhHFK8zS0WhAcGNvLTbUgA2AAWNTXAthdRbe3XxlWb8Ni7YoNH+KtGiAA) doesn't do what I'd expect on the color scale

**Comment by domoritz** [comment]
The zero property only adds 0 to the scale domain. Since there is no data when the count is zero, there is no mark, and so the scale isn't used when the count is 0.

**Comment by willium** [comment]
Yep, unless I'm missing something about the cross property, the current options aren't sufficient  to create such a chart?

**Comment by roying** [comment]
Try changing the transform bin "extent" as shown in this example.
```
  "transform": [
        {"type": "extent", "field": "IMDB Rating", "signal": "imdb_extent"},
        {
          "type": "bin",
          "field": "IMDB Rating",
          "as": ["imdb_start", "imdb_end"],
          "signal": "imdb_bins",
          "extent": {"signal": "[0, imdb_extent[1] + 0.5]"}
        },
```
[Open chart in Vega editor](https://vega.github.io/editor/#/url/vega/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykSArJQBWENgDsQAGhBMkUANZkATmwbiAJmhAB3GHTjSQOJBo01xZNAJk6aG+mgBMABhcz4NMlmduZGpEwUVABtUHEkBEN0AKCjBmUAG21Y-gQ2Yho4CBExSRkAMzZlZEw0UEwATxxokFEJEABfGUxlJHEIIpK0MJAqmu04AA9MOHEywqzErXQASQBZABEAIQACACVAiysZCC8I5PQaBA0mAH1h0fGmqQrq2qYLIwKpmZAFlY2tyyMkCB6QMdTmcIEFlBNASdzmMtABdXb7JCHSHAx4dIyXMZlVCgPZkA7aELuVZA6EjLEhACMsNWAGpVi5KAJYU1mncBuhMddJnBptp1mxsGNVgAVNilNjZL6YbZGPEE9CqIXiC7k65svr3bRo56vfmCq6i8WBSUQaWymR-AFKq4gsEQm1jC6aEDwkDypHaR0qtH-GRc7G4xHI72qq5lRoa-q1JBkFR8UZGFRqHBMSoA0l2pDgoyZmFGUOg7MOg1O-NujSqHBoApIiBwS3-UIgKBqa6wxodt3IZQKJu9CJRbQ9vtGaNeuBQCGgyqJaIhEDKSdlN0FVQIcogVIpQIoZogMatjTRUAMHCxE8gF6JQ64qBI2qtxLFXW8t6t9QRmRDJyb6AP7QhlfPkjihLMcxkNgCgKetsUZAR9yAnF3XvOdAOAt48xdSDoNgtB4P3Spf2Q-80PQdMeRAxdSxVIsIJAKCYLgOCmUIv9UNqCirz1RUaOdLQcKYliEMjTtdg4-twkiWogJaLV0ESCw4GzIwNGNJ4SODQlM19Sg6MwKQSTA3TQTYHAWX3NpLHndwg3xT10DsBwYCaN1fT-LTQNRCx-n3AAvOBVBrOs4A1QdOLHeSQEU8RlOUVT1MkTT7OREJQxM+1DPSny9MwMyLJkKyyHnOyFQwOAvB8ZoXDcnyPJSr0+Pc-zArYYLEnrMLpO0J8XzkjloqUlT-ESzdtxiXcMJ6tsv0Xdpiu0eBAmQasZAsUZlBwNhEkCWoYCgZIZACoLUFrDrQrdJAhmyHo7wA9BZIY5QsmudAmANcUk2emZWgYBs+joMj3iWNZNhlSxVgACjRWKNAASiMHamF5AAxRIGAgFzTpCmQkd5AB5UgkiQatUF+-6ZUUABREY2jQcmjosY8kJcDVSIiyDnqxbQ5wKCEVHsdr6xaQHagFZUjQlKUwe2KGYbgeHEaQZHEkJwKdtJhmAcUABhGb6rKqAKsSSHPG8TACEpFwEf8pnhjQSkxOi8gYUk7ib2m594pkFRTBezAABkxjIRxksNnaEBwU2KvNwyADYABZDNca2bgBzAgb1z9VigjZJ2KDRfI7IA)

![screenshot_vega](https://user-images.githubusercontent.com/37225856/99429790-766e8900-28d6-11eb-8081-eb46b2b8d0a4.png)

**Comment by domoritz** [comment]
@roying What is this example for? It doesn't address the issue that 0 counts don't produce a mark.
